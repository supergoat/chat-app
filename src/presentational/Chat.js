// @flow
import React, { Component } from 'react';
import type { NavigationScreenProp } from 'react-navigation';
import {
  ActivityIndicator,
  FlatList,
  Text,
  View
} from 'react-native';
import CAStyleSheet from "../common/CAStyleSheet";
import CAColors from "../common/CAColors";
import ChatBubble from "../presentational/ChatBubble";

/* Constants
============================================================================= */
const SAME_USER_MARGIN_BOTTOM = 3;
const DIFFERENT_USER_MARGIN_BOTTOM = 12;
const REFRESH_OFFSET = 20;

/** ============================================================================
<Chat
  didLoadAllMessages={boolean}
  [messages]={Array},
  onLoadMoreMessages={() => void},
  refreshing={boolean},
  viewer={string},
/>
--------------------------------------------------------------------------------
Displays the chat messages
@param { boolean } didLoadAllMessages - True if all messages have been loaded
@param { array } messages - The list of messages in the chat
@param { function } onLoadMoreMessages - Function called to load more messages
@param { boolean } refreshing - When true more messages are being fetched
@param { string } viewer - The currently logged in user
============================================================================= */
type Props = {
  didLoadAllMessages: boolean,
  messages: {
    id: string,
    text: string,
    createdAt: Date,
    createdBy: string
  }[],
  onLoadMoreMessages: () => void,
  refreshing: boolean,
  viewer: string
};
class Chat extends Component<Props> {

  static __cards__: Function

  /**
   * @description Returns a boolean if the viewer scrolled to the top
   * @param { object } nativeEvent - The scroll Event
   *
   * @returns { boolean } True if the viewer is close to the top
   */
  isCloseToTop = ({layoutMeasurement, contentOffset, contentSize}) => {
    // The offset from the top, before we should load more messages
    const paddingToTop = REFRESH_OFFSET;

    return layoutMeasurement.height + contentOffset.y >=
      contentSize.height - paddingToTop;
  };

  /**
   * @description Calculates how much margin each chatBubble should have.
   *              ChatBubbles that belong to the same author have a smaller
   *              margin between them.
   * @param { object } items - The current chat bubble
   * @param { number } index - The current index
   *
   * @returns { numbers } The chat bubble margin
   */
  getMarginBottom = (item, index) => {
    const { messages } = this.props;
    const isFirstMessage = index === 0

    let marginBottom = DIFFERENT_USER_MARGIN_BOTTOM;

    if (!isFirstMessage) {
      const previousChatBubbleAuthor = messages[index - 1].createdBy;
      const currentChatBubbleAuthor = item.createdBy;
      const isSameAuthor = currentChatBubbleAuthor === previousChatBubbleAuthor;
      if (isSameAuthor) marginBottom = SAME_USER_MARGIN_BOTTOM;
    }
    return marginBottom;
  }

  onScroll = ({nativeEvent}) => {
    const { onLoadMoreMessages, didLoadAllMessages } = this.props;

    // Return if viewer didn't scroll to top
    if (!this.isCloseToTop(nativeEvent)) return;

    if (!didLoadAllMessages) onLoadMoreMessages();;
  }

  render() {
    const {
      didLoadAllMessages,
      messages,
      refreshing,
      viewer
    } = this.props;

    return (
      <FlatList
        inverted
        style={styles.chat}
        data={messages}
        keyExtractor={item => item.id}
        onScroll={this.onScroll}
        ListFooterComponent={() => (
          <View style={styles.activityIndicator}>
            {!didLoadAllMessages &&
              <ActivityIndicator animating={refreshing} />}
          </View>
        )}
        renderItem={({ item, index }) => (
          <ChatBubble
            style={{ marginBottom: this.getMarginBottom(item, index) }}
            isAlignedLeft={item.createdBy !== viewer}
            key={item.id}
            message={item}
          />
        )}
      />
    )
  }
}

/* StyleSheet
============================================================================= */
const styles = CAStyleSheet.create({
  chat: {
    flex: 1
  },
  activityIndicator: {
    paddingVertical: 7.5
  }
});

/* Playground Cards - the different states of Chat
============================================================================= */
// Chat.__cards__ = define => {
//   define("Simple", () => {
//     return <Chat />
//   });
// };

/* Export
============================================================================= */
export default Chat;
