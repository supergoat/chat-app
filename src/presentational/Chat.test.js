// @flow
import React from 'react';
import { Text } from 'react-native';
import Chat from './Chat';
import renderer from 'react-test-renderer';
import { shallow } from 'enzyme';


describe('Chat', () => {
  it('renders correctly', () => {
    const messages = [{
      id: '1',
      text: 'Hello!',
      createdAt: new Date(1, 1, 3, 9, 5),
      createdBy: 'Panayiotis'
    }];

    const tree = renderer.create(
      <Chat
        didLoadAllMessages={false}
        messages={messages}
        onLoadMoreMessages={() => {}}
        refreshing={false}
        viewer={'Panayiotis'}
      />
    ).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('renders correctly when didLoadAllMessages is set to true', () => {
    const messages = [{
      id: '1',
      text: 'Hello!',
      createdAt: new Date(1, 1, 3, 9, 5),
      createdBy: 'Panayiotis'
    }];

    const tree = renderer.create(
      <Chat
        didLoadAllMessages={true}
        messages={messages}
        onLoadMoreMessages={() => {}}
        refreshing={false}
        viewer={'Panayiotis'}
      />
    ).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('renders correctly when refreshing is set to true', () => {
    const messages = [{
      id: '1',
      text: 'Hello!',
      createdAt: new Date(1, 1, 3, 9, 5),
      createdBy: 'Panayiotis'
    }];

    const tree = renderer.create(
      <Chat
        didLoadAllMessages={false}
        messages={messages}
        onLoadMoreMessages={() => {}}
        refreshing={true}
        viewer={'Panayiotis'}
      />
    ).toJSON();
    expect(tree).toMatchSnapshot();
  });

  describe('getMarginBottom', () => {
    it('returns 12 when its the first item', () => {
      const messages = [
        {
          id: '1',
          text: 'Hello!',
          createdAt: new Date(1, 1, 3, 9, 5),
          createdBy: 'Panayiotis'
        },
        {
          id: '2',
          text: 'Hey!',
          createdAt: new Date(1, 1, 3, 9, 5),
          createdBy: 'Alex'
        }
      ];

      const component = shallow(
        <Chat
          didLoadAllMessages={false}
          messages={messages}
          onLoadMoreMessages={() => {}}
          refreshing={true}
          viewer={'Panayiotis'}
        />
      )
      expect(component.instance().getMarginBottom(messages[0], 0)).toEqual(12);
    });

    it('returns 3 when the author of the previous item is the same', () => {
      const messages = [
        {
          id: '1',
          text: 'Hello!',
          createdAt: new Date(1, 1, 3, 9, 5),
          createdBy: 'Panayiotis'
        },
        {
          id: '2',
          text: 'How are you?',
          createdAt: new Date(1, 1, 3, 9, 5),
          createdBy: 'Panayiotis'
        }
      ];

      const component = shallow(
        <Chat
          didLoadAllMessages={false}
          messages={messages}
          onLoadMoreMessages={() => {}}
          refreshing={true}
          viewer={'Panayiotis'}
        />
      )
      expect(component.instance().getMarginBottom(messages[1], 1)).toEqual(3);
    });

    it('returns 12 when the author of the previous item is different', () => {
      const messages = [
        {
          id: '1',
          text: 'Hello!',
          createdAt: new Date(1, 1, 3, 9, 5),
          createdBy: 'Panayiotis'
        },
        {
          id: '2',
          text: 'Hey!',
          createdAt: new Date(1, 1, 3, 9, 5),
          createdBy: 'Alex'
        }
      ];

      const component = shallow(
        <Chat
          didLoadAllMessages={false}
          messages={messages}
          onLoadMoreMessages={() => {}}
          refreshing={true}
          viewer={'Panayiotis'}
        />
      )
      expect(component.instance().getMarginBottom(messages[1], 1)).toEqual(12);
    });
  });

  describe('isCloseToTop', () => {
    it('returns false when the viewer does not scroll to the top', () => {
      const messages = [{
        id: '1',
        text: 'Hello!',
        createdAt: new Date(1, 1, 3, 9, 5),
        createdBy: 'Panayiotis'
      }];

      const component = shallow(
        <Chat
          didLoadAllMessages={false}
          messages={messages}
          onLoadMoreMessages={() => {}}
          refreshing={true}
          viewer={'Panayiotis'}
        />
      );

      const nativeEvent = {
        layoutMeasurement: { height: 100 },
        contentOffset: { y: 100},
        contentSize: { height: 300}
      };
      expect(component.instance().isCloseToTop(nativeEvent)).toEqual(false);
    });

    it('returns true when the viewer scrolls to the top', () => {
      const messages = [{
        id: '1',
        text: 'Hello!',
        createdAt: new Date(1, 1, 3, 9, 5),
        createdBy: 'Panayiotis'
      }];

      const component = shallow(
        <Chat
          didLoadAllMessages={false}
          messages={messages}
          onLoadMoreMessages={() => {}}
          refreshing={true}
          viewer={'Panayiotis'}
        />
      );

      // Set the contentSize.height to a number LOWER than
      // layoutMeasurement.height + contentOffset.y
      // to simulate viewer being scrolled to the top
      const nativeEvent = {
        layoutMeasurement: { height: 100 },
        contentOffset: { y: 100},
        contentSize: { height: 190}
      };
      expect(component.instance().isCloseToTop(nativeEvent)).toEqual(true);
    });
  });

  describe('onScroll', () => {
    it('does nothing if the viewer is not scrolled to the top', () => {
      const onLoadMoreMessages = jest.fn();

      const component = shallow(
        <Chat
          didLoadAllMessages={false}
          messages={[]}
          onLoadMoreMessages={onLoadMoreMessages}
          refreshing={true}
          viewer={'Panayiotis'}
        />
      );

      component.instance().isCloseToTop = jest.fn(() => false);

      const nativeEvent = {}; // nativeEvent won't be used
      component.instance().onScroll({nativeEvent})
      expect(onLoadMoreMessages).not.toHaveBeenCalled()
    });

    it('it does nothing if didLoadAllMessages is true and viewer is scrolled to the top', () => {
      const onLoadMoreMessages = jest.fn();

      const component = shallow(
        <Chat
          didLoadAllMessages={true}
          messages={[]}
          onLoadMoreMessages={onLoadMoreMessages}
          refreshing={true}
          viewer={'Panayiotis'}
        />
      );

      component.instance().isCloseToTop = jest.fn(() => true)

      const nativeEvent = {}; // nativeEvent won't be used
      component.instance().onScroll({nativeEvent})
      expect(onLoadMoreMessages).not.toHaveBeenCalled()
    });

    it('calls onLoadMoreMessages if the viewer scrolls to the top', () => {
      const onLoadMoreMessages = jest.fn();

      const component = shallow(
        <Chat
          didLoadAllMessages={false}
          messages={[]}
          onLoadMoreMessages={onLoadMoreMessages}
          refreshing={true}
          viewer={'Panayiotis'}
        />
      );

      component.instance().isCloseToTop = jest.fn(() => true);

      const nativeEvent = {}; // nativeEvent won't be used
      component.instance().onScroll({nativeEvent})
      expect(onLoadMoreMessages).toHaveBeenCalled()
    });
  });
});
