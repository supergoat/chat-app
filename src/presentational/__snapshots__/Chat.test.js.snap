// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Chat renders correctly 1`] = `
<RCTScrollView
  ListFooterComponent={[Function]}
  data={
    Array [
      Object {
        "createdAt": 1901-02-03T09:05:00.000Z,
        "createdBy": "Panayiotis",
        "id": "1",
        "text": "Hello!",
      },
    ]
  }
  disableVirtualization={false}
  getItem={[Function]}
  getItemCount={[Function]}
  horizontal={false}
  initialNumToRender={10}
  invertStickyHeaders={true}
  inverted={true}
  keyExtractor={[Function]}
  maxToRenderPerBatch={10}
  numColumns={1}
  onContentSizeChange={[Function]}
  onEndReachedThreshold={2}
  onLayout={[Function]}
  onMomentumScrollEnd={[Function]}
  onScroll={[Function]}
  onScrollBeginDrag={[Function]}
  onScrollEndDrag={[Function]}
  renderItem={[Function]}
  scrollEventThrottle={50}
  stickyHeaderIndices={Array []}
  style={
    Array [
      Object {
        "transform": Array [
          Object {
            "scaleY": -1,
          },
        ],
      },
      Object {
        "flex": 1,
      },
    ]
  }
  updateCellsBatchingPeriod={50}
  viewabilityConfigCallbackPairs={Array []}
  windowSize={21}
>
  <View>
    <View
      onLayout={[Function]}
      style={
        Array [
          Object {
            "flexDirection": "column-reverse",
          },
          Object {
            "transform": Array [
              Object {
                "scaleY": -1,
              },
            ],
          },
        ]
      }
    >
      <View
        style={
          Array [
            Object {
              "borderRadius": 6,
              "marginBottom": 3,
              "marginHorizontal": 20,
              "maxWidth": "78%",
              "paddingHorizontal": 7,
              "paddingVertical": 5,
              "shadowColor": "#7e8890",
              "shadowOffset": Object {
                "height": 0.5,
                "width": 0,
              },
              "shadowOpacity": 1,
              "shadowRadius": 0,
            },
            Object {
              "backgroundColor": "#DFF8C8",
            },
            Object {
              "alignSelf": "flex-end",
            },
            Object {
              "marginBottom": 12,
            },
          ]
        }
      >
        <Text
          style={
            Object {
              "color": "#000000",
              "fontSize": 15,
              "lineHeight": 19,
            }
          }
        >
          Hello!
          <Text
            style={
              Object {
                "color": "transparent",
                "opacity": 0,
              }
            }
          >
            xxxxxx
          </Text>
        </Text>
        <Text
          style={
            Object {
              "bottom": 3,
              "color": "#7e8890",
              "fontSize": 11,
              "position": "absolute",
              "right": 8,
            }
          }
        >
          09:05
        </Text>
      </View>
    </View>
    <View
      onLayout={[Function]}
      style={
        Object {
          "transform": Array [
            Object {
              "scaleY": -1,
            },
          ],
        }
      }
    >
      <View
        style={
          Object {
            "paddingVertical": 7.5,
          }
        }
      >
        <ActivityIndicator
          animating={false}
          color="#999999"
          hidesWhenStopped={true}
          size="small"
        />
      </View>
    </View>
  </View>
</RCTScrollView>
`;

exports[`Chat renders correctly when didLoadAllMessages is set to true 1`] = `
<RCTScrollView
  ListFooterComponent={[Function]}
  data={
    Array [
      Object {
        "createdAt": 1901-02-03T09:05:00.000Z,
        "createdBy": "Panayiotis",
        "id": "1",
        "text": "Hello!",
      },
    ]
  }
  disableVirtualization={false}
  getItem={[Function]}
  getItemCount={[Function]}
  horizontal={false}
  initialNumToRender={10}
  invertStickyHeaders={true}
  inverted={true}
  keyExtractor={[Function]}
  maxToRenderPerBatch={10}
  numColumns={1}
  onContentSizeChange={[Function]}
  onEndReachedThreshold={2}
  onLayout={[Function]}
  onMomentumScrollEnd={[Function]}
  onScroll={[Function]}
  onScrollBeginDrag={[Function]}
  onScrollEndDrag={[Function]}
  renderItem={[Function]}
  scrollEventThrottle={50}
  stickyHeaderIndices={Array []}
  style={
    Array [
      Object {
        "transform": Array [
          Object {
            "scaleY": -1,
          },
        ],
      },
      Object {
        "flex": 1,
      },
    ]
  }
  updateCellsBatchingPeriod={50}
  viewabilityConfigCallbackPairs={Array []}
  windowSize={21}
>
  <View>
    <View
      onLayout={[Function]}
      style={
        Array [
          Object {
            "flexDirection": "column-reverse",
          },
          Object {
            "transform": Array [
              Object {
                "scaleY": -1,
              },
            ],
          },
        ]
      }
    >
      <View
        style={
          Array [
            Object {
              "borderRadius": 6,
              "marginBottom": 3,
              "marginHorizontal": 20,
              "maxWidth": "78%",
              "paddingHorizontal": 7,
              "paddingVertical": 5,
              "shadowColor": "#7e8890",
              "shadowOffset": Object {
                "height": 0.5,
                "width": 0,
              },
              "shadowOpacity": 1,
              "shadowRadius": 0,
            },
            Object {
              "backgroundColor": "#DFF8C8",
            },
            Object {
              "alignSelf": "flex-end",
            },
            Object {
              "marginBottom": 12,
            },
          ]
        }
      >
        <Text
          style={
            Object {
              "color": "#000000",
              "fontSize": 15,
              "lineHeight": 19,
            }
          }
        >
          Hello!
          <Text
            style={
              Object {
                "color": "transparent",
                "opacity": 0,
              }
            }
          >
            xxxxxx
          </Text>
        </Text>
        <Text
          style={
            Object {
              "bottom": 3,
              "color": "#7e8890",
              "fontSize": 11,
              "position": "absolute",
              "right": 8,
            }
          }
        >
          09:05
        </Text>
      </View>
    </View>
    <View
      onLayout={[Function]}
      style={
        Object {
          "transform": Array [
            Object {
              "scaleY": -1,
            },
          ],
        }
      }
    >
      <View
        style={
          Object {
            "paddingVertical": 7.5,
          }
        }
      />
    </View>
  </View>
</RCTScrollView>
`;

exports[`Chat renders correctly when refreshing is set to true 1`] = `
<RCTScrollView
  ListFooterComponent={[Function]}
  data={
    Array [
      Object {
        "createdAt": 1901-02-03T09:05:00.000Z,
        "createdBy": "Panayiotis",
        "id": "1",
        "text": "Hello!",
      },
    ]
  }
  disableVirtualization={false}
  getItem={[Function]}
  getItemCount={[Function]}
  horizontal={false}
  initialNumToRender={10}
  invertStickyHeaders={true}
  inverted={true}
  keyExtractor={[Function]}
  maxToRenderPerBatch={10}
  numColumns={1}
  onContentSizeChange={[Function]}
  onEndReachedThreshold={2}
  onLayout={[Function]}
  onMomentumScrollEnd={[Function]}
  onScroll={[Function]}
  onScrollBeginDrag={[Function]}
  onScrollEndDrag={[Function]}
  renderItem={[Function]}
  scrollEventThrottle={50}
  stickyHeaderIndices={Array []}
  style={
    Array [
      Object {
        "transform": Array [
          Object {
            "scaleY": -1,
          },
        ],
      },
      Object {
        "flex": 1,
      },
    ]
  }
  updateCellsBatchingPeriod={50}
  viewabilityConfigCallbackPairs={Array []}
  windowSize={21}
>
  <View>
    <View
      onLayout={[Function]}
      style={
        Array [
          Object {
            "flexDirection": "column-reverse",
          },
          Object {
            "transform": Array [
              Object {
                "scaleY": -1,
              },
            ],
          },
        ]
      }
    >
      <View
        style={
          Array [
            Object {
              "borderRadius": 6,
              "marginBottom": 3,
              "marginHorizontal": 20,
              "maxWidth": "78%",
              "paddingHorizontal": 7,
              "paddingVertical": 5,
              "shadowColor": "#7e8890",
              "shadowOffset": Object {
                "height": 0.5,
                "width": 0,
              },
              "shadowOpacity": 1,
              "shadowRadius": 0,
            },
            Object {
              "backgroundColor": "#DFF8C8",
            },
            Object {
              "alignSelf": "flex-end",
            },
            Object {
              "marginBottom": 12,
            },
          ]
        }
      >
        <Text
          style={
            Object {
              "color": "#000000",
              "fontSize": 15,
              "lineHeight": 19,
            }
          }
        >
          Hello!
          <Text
            style={
              Object {
                "color": "transparent",
                "opacity": 0,
              }
            }
          >
            xxxxxx
          </Text>
        </Text>
        <Text
          style={
            Object {
              "bottom": 3,
              "color": "#7e8890",
              "fontSize": 11,
              "position": "absolute",
              "right": 8,
            }
          }
        >
          09:05
        </Text>
      </View>
    </View>
    <View
      onLayout={[Function]}
      style={
        Object {
          "transform": Array [
            Object {
              "scaleY": -1,
            },
          ],
        }
      }
    >
      <View
        style={
          Object {
            "paddingVertical": 7.5,
          }
        }
      >
        <ActivityIndicator
          animating={true}
          color="#999999"
          hidesWhenStopped={true}
          size="small"
        />
      </View>
    </View>
  </View>
</RCTScrollView>
`;
